### DISCUSSION TITLE
Performance regression when moving from 4.2.0 to 5.1.5

### DISCUSSION THREAD

**Discussion by hgjunip** [Azure/azure-mobile-apps#689] [Q&A]
Hi Adrian, adding this as a discussion first before creating an issue in case I'm missing something. At our company we are moving from version 4.2.0 to version 5.1.5 and we are seeing a regression on pull performance (pulling a database that used to take 1 second now takes 7 seconds). This is related to discussion in issue #575. I'm guessing the slower performance seems to be related to the upsert that happens when pulling is now done on a per item basis instead of batching all items and then batch inserting them. 

Here is the older code for reference [Old code](https://github.com/Azure/azure-mobile-services/blob/1198c8576bf18f13f24dbc2e722bb8672c3a118c/sdk/Managed/src/Microsoft.WindowsAzure.MobileServices/Table/Sync/Queue/Actions/PullAction.cs?plain=1#L92-L136).

Here is the newer version of the pull code: [New code](https://github.com/Azure/azure-mobile-apps/blob/50fc5a6a7b6eac443bada7a41a67e5ed3fb1af03/sdk/dotnet/src/Microsoft.Datasync.Client/Offline/SyncContext.cs#L381-L395).

In my testing I have modified the new code to match the old code in terms of batch upserting the items like follows (I also had to change how the columns where calculated inside the upsert method):
```
      if (ServiceSerializer.IsDeleted(item))
      {
          //await OfflineStore.DeleteAsync(tableName, new[] { itemId }, cancellationToken).ConfigureAwait(false);
          deletes.Add((itemId, itemUpdatedAt));
      }
      else
      {
          upserts.Add((itemId, item, itemUpdatedAt));
          //await OfflineStore.UpsertAsync(tableName, new[] { item }, true, cancellationToken).ConfigureAwait(false);
      }
      itemCount++;
  }

  await OfflineStore.DeleteAsync(tableName, deletes.Select(x => x.id), cancellationToken).ConfigureAwait(false);
  await OfflineStore.UpsertAsync(tableName, upserts.Select(x => x.item), true, cancellationToken).ConfigureAwait(false);
  updatedAt = upserts.Select(x => x.updatedAt).Concat(deletes.Select(x => x.updatedAt)).Where(x => x != null).Select(x => x.Value).OrderByDescending(x => x).FirstOrDefault();

  if (updatedAt.HasValue)
  {
      await DeltaTokenStore.SetDeltaTokenAsync(tableName, queryId, updatedAt.Value, cancellationToken).ConfigureAwait(false);
      updatedAt = null; // Easy way to prevent the finally block to not write twice.
  }
  //Skip item was stored for now
  foreach (var (itemId, _, _) in upserts)
  {
      SendItemWasStoredEvent(tableName, itemId, itemCount, expectedItems);
  }
```

With these changes the pull performance is back to where it was on the old version. 

Before further investigation, I would like to understand the reason behind the change in the pull logic. Could you please provide insights into the rationale behind this change? Additionally, I would appreciate it if you could highlight any potential downsides of reverting to the old batching approach.

If you consider this change to be beneficial, I can create an issue and submit a pull request with cleaner code to implement it.

Looking forward to your feedback.

**Comment by adrianhall** [comment]
The problem is when the sync process fails in the middle.  At that point, all the previous stuff is left dangling and there is no way to recover.  This was a relatively common problem in the old code base.  This is especially important in the new code base when we can do parallel writes.

**Reply by hgjunip** [reply]
Sorry I'm not sure if I understand. Could you clarify what kind of failure and what stuff would be left dangling? You mean items would be inserted without the updatedAt being consistent with them? Also I thought since the batch upsert all happens inside a transaction a failure would just roll back everything? 
Also could you elaborate on the parallel writes?

**Reply by adrianhall** [reply]
Sure.

One of the features we added was the ability to parallelize the writes to the server.  We do this in the DatasyncClientOptions by setting the ParallelOperations value - see https://learn.microsoft.com/en-us/azure/developer/mobile-apps/azure-mobile-apps/howto/client/dotnet#paralleloperations

So what?  Well, now the ordering of the operations is not guaranteed, so we go to extra effort to ensure that the database update and the delta token store update are as close together as possible, and we do a follow-on write to the delta token store at the end of the completed push operation.  However, one of the drawbacks is that we can no longer do the bulk insert/updates.

Which is a long way around saying "there be dragons here" and to not touch this code sequence without appreciating the consequences.

That being said, when ParallelOperations == 1, we can do some short circuits.  I'll look at that next time I'm in that part of the code base.

**Reply by hgjunip** [reply]
Thanks for your time sorry to keep at it. We would like to understand a bit more of what is going on specially since our customers are using the older version and they might be experiencing the issue you mentioned and we might not know about it.

I think it might be easier with an example. Let's say we have pending push operations in a client ordered A, B, C and D. Since now we can parallel push those items the order in which they were stored in the server could be any combination of those so in this instance they were stored in the D, B, C, A order . So now when a client syncs we insert one item at a time right. So first D then B and then let's say C failed so the delta token saves the B updatedAt and then when we try again we would just pull C and A and upsert those. In the old version the client would have tried to batch insert D, B, C, A and if it failed it would just roll back all of the changes and if it succeeded then and only then it would update the delta token. Me and my colleague tried to run over the scenarios in our head and can't see clearly when the issue would arise? Are the issues related to foreign key constrains and the order in which the items get pulled on other clients or is it an issue for the client doing the pushing or is it something else entirely? 

Also you mentioned this arises from the parallel operations feature but we don't understand how having one client push on multiple threads is different from multiple clients pushing from one thread?

**Reply by amertlich** [reply]
Maybe you can help us mere peasants understand and fight these dragons ;)

> One of the features we added was the ability to parallelize the writes to the server

I understand that this means the order in which operations were saved client-side is not guaranteed to be the order they get inserted/updated and timestamped by the server. Consequently when another client pulls down, the order may be different from the original client. This is potentially a problem for objects that effectively have foreign key relationships (at least within the same table) because you're assuming that the order the client is uploading matters. Is this your thinking?

> so we go to extra effort to ensure that the database update and the delta token store update are as close together as possible

Is this so that on a pull if an individual upsert or delete fails, we have at least made some progress towards an incremental pull sync? What sort of failures are you trying to protect against here? It definitely comes with a performance penalty, dealing with individual inserts rather than a batch transaction. I'm not seeing enough value to justify this.

Are you saying that you're some how restoring the operation order on a pull sync? So the next pull may not result in a failure? -- I doubt it's this, but just asking because I'm not understanding why keeping the delta token store and the item update operations close together improves the situation.

In some scenarios this may be more dangerous if you were dealing with items that do have foreign keys, because now you just committed local database change that may require other dependent records that haven't made it down yet.  If they were in a batch/page size it seems like this scenario would still happen but at a lower probability? 

> That being said, when ParallelOperations == 1, we can do some short circuits. I'll look at that next time I'm in that part of the code base.

That would be true if you could always guarantee that the table was PUSHED with no parallel operations. However, when you're on the PULL side of this sync, there is no way to enforce or guarantee that the table WASN'T pushed with parallel operations (let alone multiple clients pushing in parallel). I think what you mean is that certain tables may inherently be insensitive to the ordering of operations. But if they are _sensitive_ to the order of operations, then they MUST be pushed without parallel operations. Perhaps you would set this up in the define table rather than the pull operation option. That way, if the table was defined to support parallel operations (where order doesn't matter), then the PULL could be performed with batch transactions.

In any case -- I also see this as being a large performance detriment and don't have a clear enough appreciation for the "relatively common problem" that existed in the old code base to feel like the performance hit is justified. Thanks in advance!

**Reply by adrianhall** [reply]
You've basically captured the essence of it.  In the v4.2.0 codebase, you had to pull and push as a single operation - they were locked against each other AND other tables, so the ordering on both pull and push was deterministic.  That is no longer the case.

I'm looking at what I can do deterministically when parallel threading is turned off to speed things up as well as ensure reliability.

**Reply by amertlich** [reply]
Thanks for the response.

In my opinion, paged or batched PUSH operations would be better than parallelizing individual push operations.  The parallelization introduces more complexity. I'd prefer to see a push operation page/batch size option rather than a number of a parallel push operations.

This would require a server side change and probably some API changes, and conflict resolutions would be more difficult.  So if you're not looking to make that big of a change then just behaving according to the old way (batched upserts/deletions on pull) -- or at least an option to configure this behavior would be a nice quick fix.

**Reply by adrianhall** [reply]
Paged or batched push operations would require significant changes to the underlying protocol.  Neither one is a "quick fix".

**Reply by amertlich** [reply]
Understood.

The quick fix I was referring to was a config option to revert to the old method. Where conditionally the PULL would behave like code hgjunip provided above.

**Reply by hgjunip** [reply]
Hi Adrian just following up. We were wondering if you think there are some changes you could make here or that we could help with if you give us a bit of guidance? Also could you point us to any unit tests that could help us verify functionality when changing this logic?

**Reply by adrianhall** [reply]
I'm happy to take a PR for these.  Open up an issue to track it.

There are plenty of unit tests that track pullasync operations, so I don't think you need to write any additional tests.  However, make sure the tests flex the new code through code coverage.

**Comment by adrianhall** [comment]
I'm just running the tests for a change.  There was a bunch of complexity here, mostly around eventing and the WriteDeltaTokenInterval.

New code works only if DatasyncClientOptions.WriteDeltaTokenInterval > 1 - in this case, we are assume you have made the trade-off between consistency and performance and sided with some level of performance.  It also gives us a middle-ground (where we write every so many records instead of at the end of the synchronization process).  Linking the writes to the delta token writes also provides a better level of consistency safety.

**Reply by hgjunip** [reply]
Thanks Adrian we really appreciate it. We had started working on this in a fork. This were the changes just for reference [PR](https://github.com/junipersys/azure-mobile-apps/pull/1). All tests were passing but we had to modify one of them because the order in which events happened changed.

**Reply by adrianhall** [reply]
I noticed the same thing. 

Check out the PR here: https://github.com/Azure/azure-mobile-apps/pull/694

I'll be merging this once the checks pass - that takes about an hour, so feel free to give it a review and leave any comments.

**Reply by hgjunip** [reply]
Sorry for the late notice but this is also related https://github.com/Azure/azure-mobile-apps/pull/694#issuecomment-1604498000
